Add functools.total_ordering

winni2k opened the pull request:
As part of my journey to port dataclasses to micropython-lib/python-stdlib (see https://github.com/orgs/micropython/discussions/13741#discussioncomment-8579351), I needed to port `functools.total_ordering` first.

I have tried to follow all the contribution guidelines, but I have not tried to optimize code size as I am completely new to micropython and would not know where to start.

This code (and test code) is copied from CPython v3.7.17 in its entirety. The PSF license appears to allow for this, but please do correct me if I am wrong.

winni2k committed:
functools: Add functools.total_ordering.

This commit is the result of copying the total_ordering code and tests
over from CPython v3.7.17.
One test is disabled because it expects builtin objects to have
attributes (__lt__, __gt__, etc.).
Another test for compatibility with pickle is also disabled because
pickle compatibility is currently broken.
Bumped package version to 0.0.8.

The functools code in CPython has the following credits:

Written by Nick Coghlan <ncoghlan at gmail.com>,
Raymond Hettinger <python at rcn.com>,
and Łukasz Langa <lukasz at langa.pl>.
  Copyright (C) 2006-2013 Python Software Foundation.
See C source code for _functools credits/copyright

This work was donated by W Winfried Kretzschmar.

Signed-off-by: W Winfried Kretzschmar <winni@warrenwk.com>

winni2k commented:
Any chance you could review this PR @mattytrentini ?

mattytrentini commented:
> Any chance you could review this PR @mattytrentini ?

Sorry, been trying to find time to do so!

Perhaps consider adding a `README.md` explaining how to run the tests? Ideally in a container but even if it's just a description it would be helpful (I mean, it looks like straightforward `unittest` code but it should be clarified). Eventually we'd like to create Github Actions to run them continuously so if we have clear instructions it'll make that task easier. 

It's also important to spell out that this code is licensed under the Python license. See the root [LICENSE](https://github.com/micropython/micropython-lib/blob/master/LICENSE) file that mentions MIT "unless explicitly stated otherwise". I _think_ an entry in the README will suffice. 

Otherwise, the code looks good! I'm curious about the failing unit test but I'll have to look into that another time...

(PS thanks for the contribution!)

winni2k commented:
Have a look now. 

- moved the PSF licensed code into a separate file, which required creating a package for functools. 
- added instructions for running tests using Docker.

projectgus reviewed (commented):
The new feature in functools look good to me, thanks @winni2k! Thanks also for your patience while someone got back to you.

Regarding the Dockerfile and README: I appreciate you providing some steps for a docker container to run tests, but this probably isn't something we want to implement differently for each module. There are already several other micropython-lib modules that include tests based on `unittest`. Most of these can be run on a host as `micropython FILENAME`, which looks like it's also the case for these tests after the dependencies are installed.

If you did want to submit another PR that adds standard instructions for "Running micropython-lib tests", and includes docker as one option for this, then that could potentially be very useful. However, suggest doing that separately to this total_ordering PR.

dpgeorge commented on python-stdlib/functools/functools/total_ordering.py:
```diff
@@ -0,0 +1,133 @@
+# infinite recursion that could occur when the operator dispatch logic
+# detects a NotImplemented result and then calls a reflected method.
+
+def _gt_from_lt(self, other, NotImplemented=NotImplemented):
```
I'm not sure why the `NotImplemented=NotImplemented` argument is there, but I doubt we need it in MicroPython.  So I suggest removing it (and all below) to keep the code small.

dpgeorge commented on python-stdlib/functools/functools/total_ordering.py:
```diff
@@ -0,0 +1,133 @@
+def _le_from_lt(self, other, NotImplemented=NotImplemented):
+    'Return a <= b.  Computed by @total_ordering from (a < b) or (a == b).'
+    op_result = self.__lt__(other)
+    return op_result or self == other
```
This is slightly more efficient written as `return self.__lt__(other) or self == other`.

dpgeorge commented on python-stdlib/functools/functools/total_ordering.py:
```diff
@@ -0,0 +1,133 @@
+def _ge_from_gt(self, other, NotImplemented=NotImplemented):
+    'Return a >= b.  Computed by @total_ordering from (a > b) or (a == b).'
+    op_result = self.__gt__(other)
+    return op_result or self == other
```
This is slightly more efficient written as `return self.__gt__(other) or self == other`.

dpgeorge commented on python-stdlib/functools/functools/total_ordering.py:
```diff
@@ -0,0 +1,133 @@
+    for opname, opfunc in _convert[root]:
+        if opname not in roots:
+# function objects have no attributes in micropython
+#            opfunc.__name__ = opname
```
Please remove this commented-out code.

dpgeorge commented on python-stdlib/functools/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM micropython/unix:v1.22
+
+RUN micropython -m mip install unittest
+RUN micropython -m mip install unittest-discover
```
As @projectgus mentioned, this is not the right place for docker files, so please remove all docker-related things from this PR.
